### DISCUSSION TITLE
Versions and backwards incompatibilities

### DISCUSSION THREAD

**Discussion by bredelings** [revbayes/revbayes#629] [Ideas]
Some improvements in revbayes might requires backward incompatibilities.  For example, making user functions faster might require changing how RevBayes matches arguments to function type signatures.  We could of course decide to never improve anything that damages backwards incompatibility, but I doubt that will happen.

So if we do decide to allow some backward incompatibilities, how might they be handled?

This isn't a new problem unique to RevBayes, so it would be interesting to consider what kind of strategies are out there.

- `meson` requires all scripts to specify a minimum meson version.  If a script uses a function or function argument that was introduced after the specified minimum version, then meson tells you that you should increase the minimum version.
- `rust` has "editions". https://doc.rust-lang.org/edition-guide/editions/  These guarantee that you can also run old code by specifying an older "edition", but allow backwards incompatible changes in newer "editions"
- `linux` requires that any functionality visible to programs outside the kernel can never change at all.  
- `C++` requires that features be marked `deprecated` for several releases before removing them.  Backwards incompatible changes are testing by compiling lots of freely-available software to assess the impact.
- others?

**Comment by bredelings** [comment]
If each revbayes script was required to be annotated with the version that it ran under, then one possibility would be to say that older scripts need older versions of revbayes.

**Comment by bjoelle** [comment]
what I do in my own software is to distinguish this by version number - minor versions keep the backward compatibility, major are not guaranteed to - I think this is more or less what BEAST2 does as well
saving the version number is a good idea but I'm not sure how to make that work in practice - maybe it would be easier to save it in the output rather than the script itself

**Reply by bjoelle** [reply]
either way I agree with previous comments that functions/arguments which have been removed should give a clear error message, not just stop working silently - this is what we did when we removed the FBD range model

**Comment by ms609** [comment]
I'm glad you started this discussion; I agree with @bjoelle's comment that version numbering is a useful way to flag to users whether they might anticipate backward incompatibility, and like your previous suggestion on Slack that RevBayes version `x.y.z.w`  might be treated as equivalent to the [semantic version](https://semver.org/) `(x.y).z.w`.

If backward-incompatible changes can be treated such that attempts to use the old behaviour trigger a message "Discontinued in version x.y.z", then this would make it possible for users to identify a version of RevBayes that can run any given historic script; if the message also specifies "replaced with method x()" then a user might readily be able to update such a script.  Of course, not all cases will be handled this easily – such as where behaviour changes.

The CMake policy approach is potentially interesting here (e.g. https://cmake.org/cmake/help/latest/policy/CMP0074.html); this allows users to clearly specify which 'old' or 'new' behaviour they intend the script to employ, whilst giving an indication as to how long 'old' behaviour is likely to be supported.

**Comment by davidcerny** [comment]
I agree with @bjoelle and @ms609 that a move toward something a little more like semantic versioning would be useful. Maybe we just need to formalize and fully commit to the `(x.y).z.a` system.

I also agree it's desirable to record what version of RevBayes a given analysis was performed with. For example, IQ-TREE prints out this beautiful detailed summary to the `.log` file associated with each analysis:

```
IQ-TREE multicore version 2.3.5 for MacOS Intel 64-bit built Jul 22 2024
Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor, Heiko Schmidt,
Dominik Schrempf, Michael Woodhams, Ly Trong Nhan, Thomas Wong

Host:    MacBook-Pro-3.local (SSE4.2, 128 GB RAM)
Command: /Users/David/Grive/Slater_Lab/iqtree-2.3.5-macOS/build/iqtree2 -g MarginoThyreo_soft_constraint.tre -p partitions.nex -B 1000 --runs 100 -T AUTO -allnni -safe --prefix MarginoThyreo
Seed:    69677 (Using SPRNG - Scalable Parallel Random Number Generator)
Time:    Tue Dec 10 01:44:24 2024
Kernel:  Safe SSE2 - auto-detect threads (16 CPU cores detected)
```

The info about the hardware the analysis was run on may seem like overkill, but it's motivated by [recent concerns](https://www.nature.com/articles/s41467-020-20005-6) about the irreproducibility of ML analyses on different CPU architectures. What I like the most, though, is the fact the random seed is always printed even if the user didn't explicitly request it or specify it. We can print an autogenerated seed too, but leave the responsibility of doing it with the user – you have to include a `printSeed()` line in your script.

Of course, we do print similar info every time a RevBayes session is started, but only to the standard output, so the responsibility of saving it again rests with the user (although HPC job schedulers like Slurm redirect stdout automatically). It'd be nice to always save it to a file instead. Like @bjoelle, I don't think writing the info into the script is workable, and it would leave us without a solution for analyses that were run interactively (which I don't think is a common workflow, but it's one we try to support). The IQ-TREE folks have it easy here, since their program always produces the same set of output files, and the user only gets to specify their prefix. We give users much more control over how many output files to produce, what to call them, or indeed whether to produce them at all.

However, there is already one file we write without asking the user: `history.txt`. So maybe the version number and related info could either go directly into it, or into a separate autogenerated file called something like `version.txt` or `analysisinfo.txt`.
